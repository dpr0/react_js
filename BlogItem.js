const { DOM, PropTypes } = React;
const { bind, assign }   = _;
const Image = props => DOM.img(props);
moment.locale('ru');

const TextBox = ({author, created_at, updated_at, text}) => (
    <div>
        <h3>{text}</h3>
        <ul>
            <li><b>author:    </b> {author}    </li>
            <li><b>created_at:</b> {created_at}</li>
            <li><b>updated_at:</b> {updated_at}</li>
        </ul>
    </div>
);
TextBox.defaultProps = {author: 'anonimus', text: 'empty'};
TextBox.propTypes = {author: PropTypes.string, dislike: PropTypes.string};


class Like extends React.Component {
    constructor(props) {
        super(props);
        this.state = { like: props.like, dislike: props.dislike };
        this.likeClick = bind(this.likeClick, this);
        this.dislikeClick = bind(this.dislikeClick, this);
    }

    likeClick() {
        this.setState({ like: this.state.like + 1 });
    }
    dislikeClick() {
        this.setState({ dislike: this.state.dislike + 1 });
    }

    render() {
        return(
            <div>
                <p>
                    <button onClick={this.likeClick}>LIKE!</button>
                    Нравится: {this.state.like}
                </p>
                <p>
                    <button onClick={this.dislikeClick}>DISLIKE!</button>
                    Не нравится: {this.state.dislike}
                </p>
            </div>
        )
    }
}
Like.defaultProps = {like: 0, dislike: 0};
Like.propTypes = {like: PropTypes.number, dislike: PropTypes.number};

class BlogItem extends React.Component {
    constructor(props) {
        super(props);
        this.meta = {
            author:     'dpro',
            created_at: moment('09-31-2017 11:11:11').format('DD.MM.YYYY, dddd, h:mm:ss'),
            updated_at: moment().format('DD.MM.YYYY, dddd, h:mm:ss'),
            like:       11,
            dislike:    null, // значение не передано
            text:       '123123',
            image: {
                src:    'http://fishki.net/picsw/092011/20/bonus/foto/049.jpg',
                height: '230',
                width:  '150',
                alt:    'qwerty'
            }
        }
    }

    render() {
        return React.createElement('div', {},
            React.createElement(TextBox, this.meta),
            React.createElement(Like,    this.meta),
            React.createElement(Image,   this.meta.image))
    }
};

ReactDOM.render(
    React.createElement('div', {},
        React.createElement(BlogItem),
        React.createElement(BlogItem),
        React.createElement(BlogItem)),
    document.getElementById('app')
);