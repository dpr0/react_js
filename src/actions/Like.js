import request from 'superagent';
import { API_ROOT } from 'constants/API';
import * as types from 'constants/actionTypes/LikeActionTypes';
const requestLike = (id) => ({
  type: types.FETCH_LIKE_REQUEST,
  id
});

const receiveLike = (response) => ({
  type: types.FETCH_LIKE_SUCCESS,
  response
});

const errorLike = () => ({
  type: types.FETCH_LIKE_ERROR
});

export function fetchLike(id) {
  return (dispatch) => {
    dispatch(requestLike());

    return request
      .post(`${API_ROOT}/post/${id}/like`)
      .end((err, response) => {
        err ? dispatch(errorLike()) : dispatch(receiveLike(response.body));
      });
  };
}
