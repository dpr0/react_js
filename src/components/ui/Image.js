import PropTypes from 'prop-types';
import DOM       from 'react-dom';

const Image = ({ src, height, width, alt }) => DOM.img({ src, height, width, alt });
Image.defaultProps = {
  src: 'http://stormweb.pro/images/cms/data/raznoe/error_404.jpg',
  height: '168',
  width: '206',
  alt: '404'
};
Image.propTypes = {src: PropTypes.string, height: PropTypes.string, width: PropTypes.string, alt: PropTypes.string};

export default Image;
