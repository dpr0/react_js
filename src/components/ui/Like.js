import React      from 'react';
import PropTypes  from 'prop-types';
import { Button } from 'semantic-ui-react';

const Like = ({ id, like, dislike, likeFunc }) => (
  <div>
    <Button
      id={`like_button_${id}`}
      color='green'
      icon='thumbs outline up'
      onClick={() => likeFunc(id, 'like')}
      label={{ basic: true, color: 'black', pointing: 'left', content: like }}
    />
    <Button
      color='red'
      icon='thumbs outline down'
      onClick={() => likeFunc(id, 'dislike')}
      label={{ basic: true, color: 'black', pointing: 'left', content: dislike }}
    />
  </div>
);
Like.defaultProps = {};
Like.propTypes = {
  id:       PropTypes.number,
  like:     PropTypes.number,
  dislike:  PropTypes.number,
  likeFunc: PropTypes.func,
};

export default Like;
