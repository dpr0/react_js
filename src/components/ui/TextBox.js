import React     from 'react';
import PropTypes from 'prop-types';
import { List }  from 'semantic-ui-react';

const TextBox = ({meta, body}) => (
  <div>
    <div className="header">
      {body}
    </div>
    <div className="meta">
      <span className="category">
        <List size="mini">
          <List.Item style={meta.author    ? {} : {display: 'none'}}>
            <List.Icon name='users' />
            <List.Content>{meta.author}</List.Content>
          </List.Item>
          <List.Item style={meta.createdAt ? {} : {display: 'none'}}>
            <List.Icon name='marker' />
            <List.Content>{meta.createdAt}</List.Content>
          </List.Item>
          <List.Item style={meta.updatedAt ? {} : {display: 'none'}}>
            <List.Icon name='linkify' />
            <List.Content>{meta.updatedAt}</List.Content>
          </List.Item>
        </List>
      </span>
    </div>
  </div>
);
TextBox.defaultProps = {body: 'empty', meta: {author: 'nobody', createdAt: 'no date'}};
TextBox.propTypes = {
  meta: PropTypes.shape({
    author:    PropTypes.string,
    createdAt: PropTypes.string,
    updatedAt: PropTypes.string
  }),
  body: PropTypes.string};

export default TextBox;
