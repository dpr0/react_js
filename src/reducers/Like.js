import { assign } from 'lodash';
import * as types from 'constants/actionTypes/LikeActionTypes';

const initialState = {
  isFetching: false,
  error: false,
  likes: 0
};

export default function (state = initialState, action) {
  switch (action) {
    case types.FETCH_LIKE_REQUEST:
      return assign({}, initialState, { isFetching: true });
    case types.FETCH_LIKE_ERROR:
      return assign({}, initialState, { error: true });
    case types.FETCH_LIKE_SUCCESS:
      return assign({}, initialState, { likes: action.likes });
    default:
      return state;
  }
}