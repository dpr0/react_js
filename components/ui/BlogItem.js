import React   from 'react';
import Like    from '../ui/Like';
import Image   from '../ui/Image';
import TextBox from '../ui/TextBox';

class BlogItem extends React.Component {
    render() {
        const { like, dislike, text, image, meta } = this.props;
        return (
            <div>
                <TextBox text={text} meta={meta} />
                <Like like={like} dislike={dislike} />
                <Image {...image} />
            </div>

        )
    }
}

export default BlogItem;
