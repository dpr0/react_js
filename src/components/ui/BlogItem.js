import React             from 'react';
import Like              from 'components/ui/Like';
import Image             from 'components/ui/Image';
import TextBox           from 'components/ui/TextBox';
import Link              from 'components/elements/Link';
import PropTypes         from 'prop-types';
import { postPath }      from 'helpers/routes';
import { Grid, Segment } from 'semantic-ui-react';

const BlogItem = ({ id, like, dislike, title, body, image, meta, likeFunc, dislikeFunc }) => (
  <div className="ui raised link card">
    <div className="content">
      <TextBox body={body} meta={meta} />
      <div className="description">
        <p>
          <Image {...image} />
          <Like
            like={like}
            dislike={dislike}
            likeFunc={() => likeFunc(id)}
            dislikeFunc={() => dislikeFunc(id)}
          />
        </p>
      </div>
    </div>
  </div>
);
BlogItem.defaultProps = {};
BlogItem.propTypes = {
  id:          PropTypes.number,
  like:        PropTypes.number,
  dislike:     PropTypes.number,
  title:       PropTypes.string,
  body:        PropTypes.string,
  image:       PropTypes.object,
  meta:        PropTypes.object,
  likeFunc:    PropTypes.func,
  dislikeFunc: PropTypes.func
};

export default BlogItem;
