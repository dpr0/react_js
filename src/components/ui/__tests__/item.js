import React            from 'react';
import PropTypes        from 'prop-types';
import ReactDom         from 'react-dom';
import { expect }       from 'chai';
import { shallow }      from 'enzyme';
import { render }       from 'enzyme';
import Link             from 'components/elements/Link';
import BlogItem         from 'components/ui/BlogItem';
import { Item }         from 'semantic-ui-react';
import Like             from 'components/ui/Like';
import Image            from 'components/ui/Image';
import TextBox          from 'components/ui/TextBox';
import Enzyme           from 'enzyme';
import Adapter          from 'enzyme-adapter-react-16';
import chai             from "chai";
import renderer         from "react-test-renderer";
import chaiJestSnapshot from "chai-jest-snapshot";
Enzyme.configure({ adapter: new Adapter() });
chai.use(chaiJestSnapshot);


describe('BlogItem', () => {
  const itemProps = {
    id:          1,
    like:        1,
    dislike:     0,
    title:       'React!',
    body:        '',
    image:       { src: '', alt: ''},
    meta:        {},
    likeFunc:    () => {}
  };

  it('renders the title', () => {
    const blogItem = render(<BlogItem {...itemProps} />);
    expect(blogItem.text()).to.contain('React!');
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDom.render(<BlogItem {...itemProps} />, div);
  });

  it('should render the title', () => {
    const item = shallow(<BlogItem {...itemProps} />);
    const header = <Link to='/post/1'>React!</Link>;
    expect(item.contains(header)).equal(true);
  });

  it('should render usual item', () => {
    const item = shallow(<BlogItem {...itemProps} />);
    expect(item).to.matchSnapshot();
  });


  it('should render usual item', () => {
    const tree = renderer.create(<BlogItem {...itemProps} />).toJSON();
    expect(tree).to.matchSnapshot();
  });
});