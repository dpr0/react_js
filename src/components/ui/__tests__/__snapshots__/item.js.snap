// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogItem should render usual item 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BlogItem
    body=""
    dislike={0}
    id={1}
    image={
      Object {
        "alt": "",
        "src": "",
      }
    }
    like={1}
    likeFunc={[Function]}
    meta={Object {}}
    title="React!"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className="ui raised link card"
      >
        <div
          className="content"
        >
          <Link
            id="post_id_1"
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/post/1"
          >
            React!
          </Link>
          <TextBox
            body=""
            meta={Object {}}
          />
          <div
            className="description"
          >
            <Image
              alt=""
              height="168"
              src=""
              width="206"
            />
            <Like
              dislike={0}
              id={1}
              like={1}
              likeFunc={[Function]}
            />
          </div>
        </div>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="content"
        >
          <Link
            id="post_id_1"
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/post/1"
          >
            React!
          </Link>
          <TextBox
            body=""
            meta={Object {}}
          />
          <div
            className="description"
          >
            <Image
              alt=""
              height="168"
              src=""
              width="206"
            />
            <Like
              dislike={0}
              id={1}
              like={1}
              likeFunc={[Function]}
            />
          </div>
        </div>,
        "className": "ui raised link card",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              id="post_id_1"
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/post/1"
            >
              React!
            </Link>,
            <TextBox
              body=""
              meta={Object {}}
            />,
            <div
              className="description"
            >
              <Image
                alt=""
                height="168"
                src=""
                width="206"
              />
              <Like
                dislike={0}
                id={1}
                like={1}
                likeFunc={[Function]}
              />
            </div>,
          ],
          "className": "content",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "React!",
              "id": "post_id_1",
              "onlyActiveOnIndex": false,
              "style": Object {},
              "to": "/post/1",
            },
            "ref": null,
            "rendered": "React!",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "body": "",
              "meta": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Image
                  alt=""
                  height="168"
                  src=""
                  width="206"
                />,
                <Like
                  dislike={0}
                  id={1}
                  like={1}
                  likeFunc={[Function]}
                />,
              ],
              "className": "description",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "alt": "",
                  "height": "168",
                  "src": "",
                  "width": "206",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "dislike": 0,
                  "id": 1,
                  "like": 1,
                  "likeFunc": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="ui raised link card"
        >
          <div
            className="content"
          >
            <Link
              id="post_id_1"
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/post/1"
            >
              React!
            </Link>
            <TextBox
              body=""
              meta={Object {}}
            />
            <div
              className="description"
            >
              <Image
                alt=""
                height="168"
                src=""
                width="206"
              />
              <Like
                dislike={0}
                id={1}
                like={1}
                likeFunc={[Function]}
              />
            </div>
          </div>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="content"
          >
            <Link
              id="post_id_1"
              onlyActiveOnIndex={false}
              style={Object {}}
              to="/post/1"
            >
              React!
            </Link>
            <TextBox
              body=""
              meta={Object {}}
            />
            <div
              className="description"
            >
              <Image
                alt=""
                height="168"
                src=""
                width="206"
              />
              <Like
                dislike={0}
                id={1}
                like={1}
                likeFunc={[Function]}
              />
            </div>
          </div>,
          "className": "ui raised link card",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                id="post_id_1"
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/post/1"
              >
                React!
              </Link>,
              <TextBox
                body=""
                meta={Object {}}
              />,
              <div
                className="description"
              >
                <Image
                  alt=""
                  height="168"
                  src=""
                  width="206"
                />
                <Like
                  dislike={0}
                  id={1}
                  like={1}
                  likeFunc={[Function]}
                />
              </div>,
            ],
            "className": "content",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "React!",
                "id": "post_id_1",
                "onlyActiveOnIndex": false,
                "style": Object {},
                "to": "/post/1",
              },
              "ref": null,
              "rendered": "React!",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "body": "",
                "meta": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Image
                    alt=""
                    height="168"
                    src=""
                    width="206"
                  />,
                  <Like
                    dislike={0}
                    id={1}
                    like={1}
                    likeFunc={[Function]}
                  />,
                ],
                "className": "description",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "alt": "",
                    "height": "168",
                    "src": "",
                    "width": "206",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "dislike": 0,
                    "id": 1,
                    "like": 1,
                    "likeFunc": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`BlogItem should render usual item 2`] = `
<div
  className="ui items"
>
  <div
    className="ui raised link card"
  >
    <div
      className="content"
    >
      <a
        id="post_id_1"
        onClick={[Function]}
        style={Object {}}
      >
        React!
      </a>
      <div>
        <div
          className="header"
        >
          
        </div>
        <div
          className="meta"
        >
          <span
            className="category"
          >
            <div
              className="ui mini list"
              role="list"
            >
              <div
                className="item"
                data-value={undefined}
                onClick={[Function]}
                role="listitem"
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                <i
                  aria-hidden="true"
                  className="users icon"
                />
                <div
                  className="content"
                />
              </div>
              <div
                className="item"
                data-value={undefined}
                onClick={[Function]}
                role="listitem"
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                <i
                  aria-hidden="true"
                  className="marker icon"
                />
                <div
                  className="content"
                />
              </div>
              <div
                className="item"
                data-value={undefined}
                onClick={[Function]}
                role="listitem"
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                <i
                  aria-hidden="true"
                  className="linkify icon"
                />
                <div
                  className="content"
                />
              </div>
            </div>
          </span>
        </div>
      </div>
      <div
        className="description"
      >
        <div>
          <img
            alt=""
            height="168"
            src=""
            width="206"
          />
        </div>
        <div>
          <div
            className="ui labeled button"
            id="like_button_1"
            onClick={[Function]}
          >
            <button
              className="ui green icon button"
              disabled={undefined}
              tabIndex={0}
            >
              <i
                aria-hidden="true"
                className="thumbs outline up icon"
              />
               
            </button>
            <div
              className="ui black left pointing basic label"
              onClick={[Function]}
            >
              1
            </div>
          </div>
          <div
            className="ui labeled button"
            onClick={[Function]}
          >
            <button
              className="ui red icon button"
              disabled={undefined}
              tabIndex={0}
            >
              <i
                aria-hidden="true"
                className="thumbs outline down icon"
              />
               
            </button>
            <div
              className="ui black left pointing basic label"
              onClick={[Function]}
            >
              0
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
